### PR TITLE
fix: add a configurable max timeout for outgoing predicate payload requests

### LINKED ISSUE #643: add a configurable max timeout for predicate requests
If chainhook is sending a POST request for a predicate into an external server but that server never responds, it leads to an indefinite lock in the thread that handles predicate evaluations, which disrupts the entire block consumption process.

We should add a configurable timeout for these requests.

### PR DESCRIPTION
**Pull request by rafaelcr**
This PR adds a new (optional) entry to the config .toml file so outgoing predicate payload HTTP POST requests can have a configurable timeout that will cancel the request if hit.

```toml
[predicates]
payload_http_request_timeout_ms = 10000
```

This new setting is useful to public-facing chainhook instances that may run into unresponsive 3rd party servers that may take forever to respond to a payload request.

Fixes #643

### CODE REVIEW

File: components/chainhook-cli/src/scan/stacks.rs [hunk 1]
```diff
@@ -338,7 +340,12 @@ pub async fn scan_stacks_chainstate_via_rocksdb_using_predicate(
             apply: hits_per_blocks,
             rollback: vec![],
         };
-        let res = match handle_stacks_hook_action(trigger, &proofs, ctx) {
+        let res = match handle_stacks_hook_action(
+            trigger,
+            &proofs,
+            &config.get_event_observer_config().predicates_config,
```
Comment (csgui) [inline components/chainhook-cli/src/scan/stacks.rs:346]: It seems more reasonable to pass the parent `EventObserverConfig` here and let the `handle_stacks_hook_action` function decide what it needs. The caller don't need to know about implementation details of the callee function.

File: components/chainhook-cli/src/scan/bitcoin.rs [hunk 2]
```diff
@@ -290,7 +290,7 @@ pub async fn execute_predicates_action<'a>(
             gather_proofs(&trigger, &mut proofs, config, ctx);
         }
         let predicate_uuid = &trigger.chainhook.uuid;
-        match handle_bitcoin_hook_action(trigger, &proofs) {
+        match handle_bitcoin_hook_action(trigger, &proofs, &config.predicates_config) {
```
Comment (csgui) [inline components/chainhook-cli/src/scan/bitcoin.rs:293]: It seems more reasonable to pass the parent `EventObserverConfig` here and let the `handle_bitcoin_hook_action` function decide what it needs. The caller don't need to know about implementation details of the callee function.

File: components/chainhook-cli/src/scan/stacks.rs [hunk 3]
```diff
@@ -525,7 +534,12 @@ pub async fn scan_stacks_chainstate_via_csv_using_predicate(
             apply: hits_per_blocks,
             rollback: vec![],
         };
-        match handle_stacks_hook_action(trigger, &proofs, ctx) {
+        match handle_stacks_hook_action(
+            trigger,
+            &proofs,
+            &config.get_event_observer_config().predicates_config,
```
Comment (csgui) [inline components/chainhook-cli/src/scan/stacks.rs:540]: Same comment above.

File: components/chainhook-sdk/src/chainhooks/stacks/mod.rs [hunk 4]
```diff
@@ -1325,11 +1327,16 @@ pub fn serialize_stacks_payload_to_json<'a>(
...
     match &trigger.chainhook.action {
         HookAction::HttpPost(http) => {
-            let client = Client::builder()
+            let mut client_builder = Client::builder();
+            if let Some(timeout) = config.payload_http_request_timeout_ms {
+                client_builder = client_builder.timeout(Duration::from_millis(timeout));
+            }
+            let client = client_builder
                 .build()
                 .map_err(|e| format!("unable to build http client: {}", e))?;
```
Comment (csgui) [inline components/chainhook-sdk/src/chainhooks/stacks/mod.rs:1341]: What you think in exchange mutability for method chaining? A mutable `client` instance is less robust than an immutable one, IMO.

And with `unwrap_or_else` a default timeout can be set.

```rust
let client = Client::builder()
    .timeout(config.payload_http_request_timeout_ms
        .map(Duration::from_millis)
        .unwrap_or_else(|| Duration::from_secs(30))) // <<<--- Default timeout
    .build()
    .map_err(|e| format!("unable to build http client: {}", e))?;
```

**Review by csgui — CHANGES_REQUESTED** [review]
Added some comments and changes as suggestion.

File: components/chainhook-sdk/src/chainhooks/stacks/mod.rs (hunk 4 above)
Comment (rafaelcr) [reply components/chainhook-sdk/src/chainhooks/stacks/mod.rs]: I agree with your suggestion, the problem is that we want the default to be "no timeout" and there's no way to set that via the `ClientBuilder`. I'm not sure why they don't have a method to provide an `Optional<Duration>`

Comment (csgui) [reply components/chainhook-sdk/src/chainhooks/stacks/mod.rs]: Ok. 👍

**Review by csgui — APPROVED** [review]
Code LGTM!

**Pull request merged**